pkg/service/service_test.go

```diff
@@ -579,3 +581,106 @@ func TestGetAlmExample(t *testing.T) {
 		})
 	}
 }
+
+func TestListSucceededClusterServiceVersions(t *testing.T) {
+
+	tests := []struct {
+		name              string
+		isCSVsupported    bool
+		isCSVsupportedErr error
+		list              *olm.ClusterServiceVersionList
+		listErr           error
+		expectedList      *olm.ClusterServiceVersionList
+		expectedErr       bool
+	}{
+		{
+			name:              "error getting supported csv",
+			isCSVsupported:    false,
+			isCSVsupportedErr: errors.New("an error"),
+			expectedList:      &olm.ClusterServiceVersionList{},
+			expectedErr:       true,
+		},
+		{
+			name:              "non supported csv",
+			isCSVsupported:    false,
+			isCSVsupportedErr: nil,
+			expectedList:      &olm.ClusterServiceVersionList{},
+			expectedErr:       false,
+		},
+		{
+			name:              "error getting list",
+			isCSVsupported:    true,
+			isCSVsupportedErr: nil,
+			list:              nil,
+			listErr:           errors.New("an error"),
+			expectedList:      &olm.ClusterServiceVersionList{},
+			expectedErr:       true,
+		},
+		{
+			name:              "supported csv, empty list",
+			isCSVsupported:    true,
+			isCSVsupportedErr: nil,
+			list:              &olm.ClusterServiceVersionList{},
+			expectedList:      &olm.ClusterServiceVersionList{},
+			expectedErr:       false,
+		},
+		{
+			name:              "supported csv, return succeeded only",
+			isCSVsupported:    true,
+			isCSVsupportedErr: nil,
+			list: &olm.ClusterServiceVersionList{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "a kind",
+					APIVersion: "a version",
+				},
+				Items: []olm.ClusterServiceVersion{
+					{
+						Status: olm.ClusterServiceVersionStatus{
+							Phase: "Succeeded",
+						},
+					},
+					{
+						Status: olm.ClusterServiceVersionStatus{
+							Phase: "",
+						},
+					},
+					{
+						Status: olm.ClusterServiceVersionStatus{
+							Phase: "other phase",
+						},
+					},
+				},
+			},
+			expectedList: &olm.ClusterServiceVersionList{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "a kind",
+					APIVersion: "a version",
+				},
+				Items: []olm.ClusterServiceVersion{
+					{
+						Status: olm.ClusterServiceVersionStatus{
+							Phase: "Succeeded",
+						},
+					},
+				},
+			},
+			expectedErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			kc := kclient.NewMockClientInterface(ctrl)
+			kc.EXPECT().IsCSVSupported().Return(tt.isCSVsupported, tt.isCSVsupportedErr).AnyTimes()
+			kc.EXPECT().ListClusterServiceVersions().Return(tt.list, tt.listErr).AnyTimes()
+			got, gotErr := ListSucceededClusterServiceVersions(kc)
+			if gotErr != nil != tt.expectedErr {
```

Comment:
That's the first time I am seeing a comparison like this. How does it work? Is it comparing that neither of `gotErr`, `nil`, `tt.expectedErr` are same?

Reply:
It can be read as:
```
if (gotErr != nil) != tt.expectedErr {
```